### DISCUSSION TITLE
Framebuffer - Sample uno app does nothing

### DISCUSSION THREAD

**Discussion by dougcooper** [unoplatform/uno#9473] [Q&A]
Im building a custom linux system for x86_64 that uses framebuffer and when I run the [sample app](https://platform.uno/docs/articles/features/using-linux-framebuffer.html) there is no output. When I press enter the app exits and the screen goes white except for the terminal text. I have roughly verified my setup using some directfb and qt example apps that output to the screen and accept input as expected. Please let me know if you have any suggestions to get this working.

**Comment by jeromelaban** [comment]
We've made recent changes to the Framebuffer support to the preview builds of Uno 4.5. Can you try again by updating to the pre-release packages ? Also, what does the console tell you? Any error messages?

**Reply by dougcooper** [reply]
Thanks for the timely response. Ive upgraded the project to the latest release:
```
<PackageReference Include="Uno.WinUI.Skia.Linux.FrameBuffer" Version="4.5.0-dev.471" />
```

There is no output from the app on the console. Any troubleshooting tips are welcome!

**Reply by workgroupengineering** [reply]
see https://github.com/unoplatform/uno/issues/9422#issuecomment-1207583475

**Reply by jeromelaban** [reply]
If there's no logging, it's also likely that you're building the release configuration (e.g. `-c Release` on the command line). Try building using the debug configuration.

**Reply by dougcooper** [reply]
I updated code per the reference comment to increase verbosity. Here is what the output looks like on the physical device. ![image](https://user-images.githubusercontent.com/17769091/184364170-375de3da-ca0d-4395-96c9-ec3c13e32ed7.jpeg)

**Reply by dougcooper** [reply]
fbset output

```
mode "1920x1080"
    geometry 1920 1080 1920 1080 32
    timings 0 0 0 0 0 0 0
    accel true
    rgba 8/16,8/8,8/0,0/0
endmode
```

**Reply by jeromelaban** [reply]
Thanks. It looks like you're getting the fonts issue https://github.com/unoplatform/uno/issues/6441 we still have not solved yet, where the default font cannot be found by Skia for some reason.

Which distribution are you using? It may be that there's a missing font system package that causes this particular problem.

**Reply by dougcooper** [reply]
its a custom distro...ill see if i can resolve the font issue and retest.

**Reply by jeromelaban** [reply]
Thanks. Otherwise, you can specify an explicit `FontFamily` on the TextBlock, like `Courier New` or similar. If you find which package is needed for the fonts to work, it'll help, as we're still not able to reproduce this on our end.

**Reply by dougcooper** [reply]
I found the font, but am still experiencing an issue with the text and I'm not sure what to make of it. 

![image](https://user-images.githubusercontent.com/17769091/184411314-46ab64d7-cb94-4d52-893d-00f9f1340572.jpeg)

**Reply by dougcooper** [reply]
Font was available [here](https://github.com/mrbvrz/segoe-ui-linux) for non standard distros or exists as a package called `ttf-mscorefonts-installer` for debian based distros. I think putting a note in the [docs](https://platform.uno/docs/articles/features/using-linux-framebuffer.html) would be a welcome addition.

**Reply by jeromelaban** [reply]
For the text at the top, it's working properly, it's only the terminal overriding the framebuffer as the scrolls up. When testing this, it's generally best to run the app from separate terminal (through SSH for instance) and disable the cursor for the terminal on screen.

**Reply by jeromelaban** [reply]
Also, @dougcooper, out of curiosity, what led you to know about Uno's framebuffer support?

**Reply by dougcooper** [reply]
> For the text at the top, it's working properly, it's only the terminal overriding the framebuffer as the scrolls up. When testing this, it's generally best to run the app from separate terminal (through SSH for instance) and disable the cursor for the terminal on screen.

Ah I see thanks. I'll look into disabling this. Some sample apps I tested written c/c++ must be doing this automatically.

**Reply by dougcooper** [reply]
> Also, @dougcooper, out of curiosity, what led you to know about Uno's framebuffer support?

A colleague of mine is writing the app and uses WPF but was keen on giving UNO a try due to it's cross-os compatible UI capabilities. Our goal is running it as a kiosk app. I'm currently building up the embedded system side of things, and am evaluating both framebuffer and gtk+3 uno interfaces. For gtk I'm using wayland/Weston with a kiosk shell (wip).

**Reply by jeromelaban** [reply]
Thanks for the details, it's always fun to know what developers do with our bits!

**Reply by dougcooper** [reply]
@jeromelaban concerning the framebuffer console outputting text on top of my app - could this have something to do with an absence of refresh rate from the uno app? Im struggling to disable the console without disabling graphics altogether and am still wondering why other apps using the directfb api do not have an issue with the console text.

**Reply by jeromelaban** [reply]
@dougcooper Indeed, the app only refreshes the framebuffer when necessary. In practice, you could remove the logging done through either lines:
https://github.com/unoplatform/uno/blob/6ceaab20dff19bc9c9cffa04bf1c76655a5c22e6/src/SolutionTemplate/UnoSolutionTemplate.WinUI.net6/Shared/App.xaml.cs#L133-L137

to use something like https://github.com/serilog/serilog-extensions-logging, and enable a file logger that you could `tail` on another terminal or view afterward. Otherwise, if you can also disable logging entirely and: https://github.com/unoplatform/uno/blob/6ceaab20dff19bc9c9cffa04bf1c76655a5c22e6/src/SolutionTemplate/UnoSolutionTemplate.WinUI.net6/Skia.Linux.FrameBuffer/Program.cs#L12 
should theoretically be enough to make the cursor disappear.

**Reply by dougcooper** [reply]
No joy. `Console.CursorVisible = false;` was already set and i commented out the following but still receive a cursor:

https://github.com/unoplatform/uno/blob/6ceaab20dff19bc9c9cffa04bf1c76655a5c22e6/src/SolutionTemplate/UnoSolutionTemplate.WinUI.net6/Shared/App.xaml.cs#L24

**Reply by jeromelaban** [reply]
Interesting. It could be that different terminals don't handle this the same way. Just to be sure, this does not work, if you create a simple console app on your system:

```csharp
Console.CursorVisible = true;
Console.WriteLine("Cursor visible");
System.Threading.Thread.Sleep(3000);
Console.CursorVisible = false;
Console.WriteLine("Cursor hidden");
System.Threading.Thread.Sleep(3000);
Console.CursorVisible = true;
```

where the cursor is always visible?

**Reply by dougcooper** [reply]
yes that code works as expected. i cleaned and rebuilt my uno app then retested. im actually testing it two ways:

1. remotely via ssh - in this case the app displays as expected on the physical screen but the terminal cursor on the physical screen is still present....**the cursor in my ssh session is hidden**.
2. locally - in this case the app does not render until i get input from the keyboard (any key) which in turn causes the keyboard character to be rendered, but **the local cursor is in fact disabled**. 

so now it seems like there are two other issues to resolve:
* how to render the app locally without first needing keyboard input
* how to intercept keypresses so the console does not display them

**Reply by jeromelaban** [reply]
Ok, that makes sense. Currently, the keyboard support was originally developed for touch screen devices, with limited input.

We'll make a small change to get this:
https://github.com/unoplatform/uno/blob/6ceaab20dff19bc9c9cffa04bf1c76655a5c22e6/src/Uno.UI.Runtime.Skia.Linux.FrameBuffer/FramebufferHost.cs#L56

to be something similar to this:
```csharp
while (true)
{
	var key = Console.ReadKey(true);
	if (key.Key == ConsoleKey.Escape)
	{
		break;
	}
}
```

You'll likely need to handle `Window.KeyDown` events to terminate the application, but we'll change the template to make this easy.

**Reply by dougcooper** [reply]
sounds good. i should also add that enter key exits the application without any additional logic in the app.

**Reply by jeromelaban** [reply]
Correct, this is what the `ReadLine` statement at the top does.

**Reply by dougcooper** [reply]
> Correct, this is what the `ReadLine` statement at the top does.

oof. 👎 "fails code test" 👎

can you make that escape key configurable when calling Run(...) so that it can be a different key or disabled?

**Reply by jeromelaban** [reply]
That's what I've been mentioning here https://github.com/unoplatform/uno/discussions/9473#discussioncomment-3400967 that we'll change.

**Reply by dougcooper** [reply]
Ok. Sorry for the confusion. I'll wait and see what you come up with.

**Reply by jeromelaban** [reply]
The update for the framebuffer has been pushed in the prerelease bits. See [the preview documentation](https://github.com/unoplatform/uno/blob/master/doc/articles/features/using-linux-framebuffer.md#get-started-with-the-framebuffer) on how to handle graceful exit of the app.
